moby/moby
# Revendor Microsoft/go-winio for 8gB file fix when importing or committing image layers

**- What I did**

Revendored https://github.com/microsoft/go-winio to pull https://github.com/microsoft/go-winio/pull/175, which replaces a bundled fork of archive/tar for Go 1.6 with use of vanilla archive/tar.

There's more context on this issue on https://github.com/microsoft/Windows-Containers/issues/43.

Fixes: #40444

**- How I did it**

* Implemented the test `github.com/docker/docker/integration/build TestBuildWithHugeFile` for Windows, to reproduce the issue.
* Changed the revision in vendor.conf for _github.com/Microsoft/go-winio_ to the master revision containing https://github.com/microsoft/go-winio/pull/175 (Full diff: https://github.com/microsoft/go-winio/compare/6c72808b55902eae4c5943626030429ff20f3b63...5b44b70ab3ab4d291a7c1d28afe7b4afeced0ed4, thank you @cpuguy83 for the reminder)
* Ran `vndr -whitelist=^archive/tar github.com/Microsoft/go-winio` (per the note at the end of vendor.conf)
* Ran `git checkout vendor/archive/tar` to revert vndr's deletion of _vendor/archive/tar_.
* Changed the one user of `github.com/Microsoft/go-winio/archive/tar` to use `archive/tar` instead.

**- How to verify it**

The existing unit test `github.com/docker/docker/integration/build TestBuildWithHugeFile` has been implemented for Windows Containers, and reproduces the issue.

For manual reproduction, see #40444 or the below procedure.

1. Given an 8gB file, e.g.,
```powershell
$writer = [System.IO.File]::OpenWrite('8gfile')
$random = new-object Random
$blockSize = 1073741824
$bytes = new-object byte[] $blockSize
for ($i=0; $i -lt 8; $i++)
{
 $random.NextBytes($bytes)
 $writer.Write($bytes, 0, $blockSize)
}
$writer.Close()
```
2. And the following Dockerfile
```
FROM mcr.microsoft.com/windows/nanoserver:1909
COPY 8gfile 8gfile
```
3. `docker build .`

**- Description for the changelog**

> Fix import or commit of Windows Container image layers containing files larger than 8gB by updating vendored go-winio library

**- A picture of a cute animal (not mandatory but encouraged)**

[![A platypus being photogenic](https://i.imgur.com/rALUXtZ.jpg)](https://i.imgur.com/rALUXtZ)
